### PR TITLE
feat: create endpoint for changing user password

### PR DESCRIPTION
**Pull request by trevorjob**
## Description

This pull request introduces an endpoint for changing the password of authenticated users.

​The changes include:
- **Endpoint Addition**: Added PATCH ``` /api/v1/users/me/password  ``` for changing user passwords.
- **Authentication**: Ensured the endpoint is secured and only accessible to authenticated users.
- **Validation**: Implemented checks for validating old passwords and ensuring new passwords meet security criteria.
- **Error Handling**: Added appropriate error messages for invalid requests and authentication failures.
- **Testing**: Added unit and integration tests to ensure functionality and edge cases are covered.

## Related Issue (Link to issue ticket)
[ticket](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/284)

## Motivation and Context

This feature is essential to allow users to securely change their passwords.

## How Has This Been Tested?

- **Unit Tests**: Added tests to validate the endpoint responsible for changing user passwords.
- **Integration Tests**: Included end-to-end tests to ensure proper functionality and security.
**Tests Include:**
     * Changing a password with valid old and new passwords.
     * Ensuring the old password is correctly validated.
     * Handling validation errors and authentication failures.
​

## Screenshots (if appropriate - Postman, etc):
- **SUCCESSFULLY CHANGED PASSWORD**
successfully_changed

- **WITHOUT AUTHENTICATION HEADER**
without_auth_header

- **MISSING BODY ARGUMENTS**
missing_arguments

- **WRONG OLD PASSWORD**
with_wrong_password

### **Docs**
doc-top

doc-down

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
      ​
## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: api/v1/routes/user.py [hunk 1]
```diff
@@ -60,3 +61,19 @@ async def reactivate_account(request: Request, db: Session = Depends(get_db)):
         status_code=200,
         message='User reactivation successful',
     )
+
+@user.patch("/current-user/change-password", status_code=200)
```
Comment (zxenonx) [inline api/v1/routes/user.py:65]:
```suggestion
@user.patch("/me/password", status_code=200)
```
or 
```suggestion
@user.patch("/:id/password", status_code=200)
```

File: api/v1/services/user.py [hunk 2]
```diff
@@ -297,7 +297,28 @@ def reactivate_user(self, db: Session, token: str):
...
+        if not self.verify_password(old_password, user.password):
+            raise HTTPException(status_code=400, detail="Incorrect old password")
+
+        # # Update password (you need to implement this logic)
+        user.password = self.hash_password(
+            new_password
+        )  # Make sure to hash the new password
+
+        # # Commit changes to deactivate the user
+        db.commit()
```
Comment (zxenonx) [inline api/v1/services/user.py:320]: remove the comments

File: tests/v1/change_user_password_test.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+from uuid_extensions import uuid7
+
+from api.db.database import get_db
+from api.v1.models.user import User
+from api.v1.services.user import user_service
+from main import app
+
+client = TestClient(app)
+LOGIN_ENDPOINT = "api/v1/auth/login"
+CHANGE_PWD_ENDPOINT = "/api/v1/users/current-user/change-password"
```
Comment (zxenonx) [inline tests/v1/change_user_password_test.py:23]: update the route accordingly

**Review by zxenonx — CHANGES_REQUESTED** [review]
Thanks @trevorjob 
I left some comments

**Comment by trevorjob** [comment]
@zxenonx done

File: api/v1/routes/user.py [hunk 4]
```diff
@@ -6,8 +6,9 @@
 
 from api.utils.success_response import success_response
 from api.v1.models.user import User
-from api.v1.schemas.user import DeactivateUserSchema, UserBase
+from api.v1.schemas.user import DeactivateUserSchema, UserBase, ChangePasswordSchema
 from api.db.database import get_db
+# from api.utils.dependencies import get_current_user
```
Comment (Goketech) [inline api/v1/routes/user.py:11]: you added this comment

**Review by Goketech — CHANGES_REQUESTED** [review]

**Comment by trevorjob** [comment]
@Goketech i've removed it

**Review by Goketech — APPROVED** [review]

**Review by Goketech — DISMISSED** [review]

**Review by Goketech — APPROVED** [review]

**Review by Goketech — APPROVED** [review]

**Comment by Goketech** [comment]
@bolexs

**Comment by bolexs** [comment]
resolve conflicts

**Comment by trevorjob** [comment]
@bolexs done

**Review by Goketech — APPROVED** [review]

**Review by joboy-dev — APPROVED** [review]

**Review by joboy-dev — DISMISSED** [review]

**Comment by joboy-dev** [comment]
@bolexs

File: api/v1/services/user.py [hunk 5]
```diff
@@ -18,64 +18,67 @@
...
 
         return query.all()
 
-    
     def fetch(self, db: Session, id):
-        '''Fetches a user by their id'''
+        """Fetches a user by their id"""
 
         user = check_model_existence(db, User, id)
         return user
```
Comment (zxenonx) [inline api/v1/services/user.py:45]:
```suggestion
        return check_model_existence(db, User, id)
        
```

File: api/v1/services/user.py [hunk 6]
```diff
@@ -110,218 +113,236 @@ def create_admin(self, db:Session , schema : user.UserCreate):
...
-            'type': 'access'
-        }
+        """Function to create access token"""
+
+        expires = dt.datetime.now(dt.timezone.utc) + dt.timedelta(
+            minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES
+        )
+        data = {"user_id": user_id, "exp": expires, "type": "access"}
         encoded_jwt = jwt.encode(data, settings.SECRET_KEY, settings.ALGORITHM)
         return encoded_jwt
```
Comment (zxenonx) [inline api/v1/services/user.py:169]:
```suggestion
        return  jwt.encode(data, settings.SECRET_KEY, settings.ALGORITHM)

```

File: api/v1/services/user.py [hunk 7]
```diff
@@ -110,218 +113,236 @@ def create_admin(self, db:Session , schema : user.UserCreate):
...
+        """Function to create access token"""
+
+        expires = dt.datetime.now(dt.timezone.utc) + dt.timedelta(
+            days=settings.JWT_REFRESH_EXPIRY
+        )
+
+        data = {"user_id": user_id, "exp": expires, "type": "refresh"}
+
         encoded_jwt = jwt.encode(data, settings.SECRET_KEY, settings.ALGORITHM)
         return encoded_jwt
```
Comment (zxenonx) [inline api/v1/services/user.py:181]:
```suggestion
        return jwt.encode(data, settings.SECRET_KEY, settings.ALGORITHM)

```

File: api/v1/services/user.py [hunk 8]
```diff
@@ -110,218 +113,236 @@ def create_admin(self, db:Session , schema : user.UserCreate):
...
-        
+
         if token:
             access = self.create_access_token(user_id=token.id)
             refresh = self.create_refresh_token(user_id=token.id)
-            
+
             return access, refresh
         else:
             pass
```
Comment (zxenonx) [inline api/v1/services/user.py:244]:
```suggestion

```

File: api/v1/services/user.py [hunk 9]
```diff
@@ -110,218 +113,236 @@ def create_admin(self, db:Session , schema : user.UserCreate):
...
         )
-        
+
         token = self.verify_access_token(access_token, credentials_exception)
 
-        user =  db.query(User).filter(User.id == token.id).first()
-        
+        user = db.query(User).filter(User.id == token.id).first()
+
         return user
```
Comment (zxenonx) [inline api/v1/services/user.py:261]:
```suggestion
        return db.query(User).filter(User.id == token.id).first()
```

File: api/v1/services/user.py [hunk 10]
```diff
@@ -110,218 +113,236 @@ def create_admin(self, db:Session , schema : user.UserCreate):
...
-        reactivation_link = f'https://{request.url.hostname}/api/v1/users/accounts/reactivate?token={token}'
+        reactivation_link = f"https://{request.url.hostname}/api/v1/users/accounts/reactivate?token={token}"
 
         # mail_service.send_mail(
-        #     to=user.email, 
-        #     subject='Account deactivation', 
+        #     to=user.email,
+        #     subject='Account deactivation',
         #     body=f'Hello, {user.first_name},\n\nYour account has been deactivated successfully.\nTo reactivate your account if this was a mistake, please click the link below:\n{request.url.hostname}/api/users/accounts/reactivate?token={token}\n\nThis link expires after 15 minutes.'
         # )
```
Comment (zxenonx) [inline api/v1/services/user.py:289]: to remove

File: api/v1/services/user.py [hunk 11]
```diff
@@ -110,218 +113,236 @@ def create_admin(self, db:Session , schema : user.UserCreate):
...
         # Send aail to user
         # mail_service.send_mail(
-        #     to=user.email, 
-        #     subject='Account reactivation', 
+        #     to=user.email,
+        #     subject='Account reactivation',
         #     body=f'Hello, {user.first_name},\n\nYour account has been reactivated successfully'
         # )
 
         # Commit changes to deactivate the user
```
Comment (zxenonx) [inline api/v1/services/user.py:325]: to remove

File: tests/v1/change_user_password_test.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    mock_db_session.query.return_value.filter.return_value.first.return_value = (
+        mock_user
+    )
+
+    # mock_db_session.return_value.__enter__.return_value = mock_user
+    # mock_user_service.hash_password.return_value = "hashed_password"
+    # mock_db_session.add.return_value = None
+    # mock_db_session.commit.return_value = None
+    # mock_db_session.refresh.return_value = None
+
```
Comment (zxenonx) [inline tests/v1/change_user_password_test.py:69]: to delete

**Review by zxenonx — CHANGES_REQUESTED** [review]
Thanks @trevorjob 
I left some comments

**Comment by trevorjob** [comment]
@zxenonx i am done

**Review by Goketech — APPROVED** [review]

**Review by zxenonx — APPROVED** [review]
Thanks @trevorjob
